<b>Привет!</b>

Меня зовут Харочкина Екатерина, и я буду проверять твой проект. Давай общаться на «ты», но если это неудобно, пожалуйста, сообщи. Моя основная цель — не указать на совершенные тобою ошибки, а поделиться своим опытом и помочь тебе совершенствоваться как профессионалу.

Ты подготовила замечательный проект, но есть аспекты, которые еще можно улучшить. Я буду использовать комментарии с цветовой разметкой. Пожалуйста, **не перемещай, не изменяй и не удаляй их**. Это поможет выполнить повторную проверку твоего проекта оперативнее.

<div class="alert alert-danger">
<b>Комментарий ревьюера №1</b> ❌
    
Таким комментарием выделены самые важные замечания. Без их отработки проект не будет принят.
</div>

<div class="alert alert-warning">
<b>Комментарий ревьюера №1</b> ⚠️
    
Таким комментарием выделены небольшие замечания и рекомендации, которые будет хорошо учесть.
</div>

<div class="alert alert-success">
<b>Комментарий ревьюера №1</b> ✅
    
Таким комментарием выделены удачные решения🙂
</div>

Свои комментарии ты можешь помечать синим цветом.

<div class="alert alert-info">
<b>Комментарий студента №1</b>
    
Так ты можешь оформить свой комментарий
</div>

Давай работать над проектом в диалоге: если ты **что-то меняешь** в проекте по моим рекомендациям — **пиши об этом**. После того, как ты все доработаешь, я смогу принять твой проект!) 

# Исследование надежности заемщиков


<div class="alert alert-success">
<b>Комментарий ревьюера №1</b> ✅
    
Можно оформить описание проекта
</div>

Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


<div class="alert alert-success">
<b>Комментарий ревьюера №1</b> ✅
    
Разделы работы можно заканчивать промежуточными выводами
</div>

## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

<div class="alert alert-danger">
<b>Комментарий ревьюера №1</b> ❌ Замечание по всему разделу ответов на вопросы 
    
В каждом вопросе нужно вывести сводную. Здесь можно воспользоваться `pivot_table()`.
    
Чтобы найти долю должников, нужно количество должников поделить на общее число заемщиков. Эту долю можно найти с помощью функции агрегации `mean`
    
Попробуй получить сводную следующего вида. В индексе столбец, который исследуем (в первом вопросе это количество детей), а столбцы: 'Всего кредитополучателей', 'Всего должников', 'Доля должников'. Учитывай, что использовать параметр columns необязательно, а в параметр aggfunc можно указывать несколько функций агрегации сразу.

Также могу посоветовать переименовать столбцы и написать функцию для построения сводных во всех вопросах.
</div>

<div class="alert alert-info">
<b>Комментарий студента №1</b>
    
Добавила метод перед началом задач, который будет возвращать сводную таблицу. Добавила в каждое задание ещё один блок с кодом, который пометила как вариант 2
</div>

In [22]:
# Универсальный метод для дальнейших задач анализа,который будет возвращать красивую сводную таблицу
def form_pivot_table(data, group_by_value):
    # Формируем сводную таблицу
    temp_for_pivot_table = data.pivot_table(index = [group_by_value], values = 'debt', aggfunc = {'count','sum','mean'})
    # Переименовываем столбцы для удобства
    temp_for_pivot_table = temp_for_pivot_table.rename(columns={'count':'Всего кредитополучателей','sum':'Всего должников','mean':'Доля должников'}) 
    # Сортируем по значениям, по которым группируем
    temp_for_pivot_table = temp_for_pivot_table.sort_values(by=group_by_value) 
    return temp_for_pivot_table[['Всего кредитополучателей','Всего должников', 'Доля должников']]

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [23]:
# ************ 1 вариант решения (до ревью) ************

# поймём, какие категории по кол-ву детей есть
children_quantitys = sorted(data['children'].unique())
# print(children_quantitys) 
# Т.к. вариантов не много, можем не объединять разное кол-во детей в отдельные категории
# Выведем информацию в виде сводной таблицы, где:
# Группировка будет по кол-ву детей: index=['children']
# Группируем значения по значениям задолженности: columns='debt'
# В качестве значений будет кол-во клиентов*: values = 'purpose'
# Кол-во клиентов находим через аггрегирующую функцию count: aggfunc='count'
data_pivot = data.pivot_table(index=['children'], columns='debt', values = 'purpose', aggfunc='count')
# print(data_pivot) # посмотрим, что получилось
# *В values можем положить любой столбец кроме children и debt, это не повлияет на результат

# Переберём children_quantitys и посчитаем процент заложенностей по каждому:
for children_quantity in children_quantitys:
    result = data_pivot[1][children_quantity] / (data_pivot[1][children_quantity] + data_pivot[0][children_quantity])
    print(f'Кол-во детей: {children_quantity}, процент просрочек: {result*100:.2f}%')

Кол-во детей: 0, процент просрочек: 7.54%
Кол-во детей: 1, процент просрочек: 9.23%
Кол-во детей: 2, процент просрочек: 9.45%
Кол-во детей: 3, процент просрочек: 8.18%
Кол-во детей: 4, процент просрочек: 9.76%
Кол-во детей: 5, процент просрочек: nan%


In [24]:
# ************ 2 вариант решения (после 1 ревью) ************

form_pivot_table(data, 'children')

Всего кредитополучателей  Всего должников  Доля должников
children                                                           
0                          14091.0           1063.0        0.075438
1                           4808.0            444.0        0.092346
2                           2052.0            194.0        0.094542
3                            330.0             27.0        0.081818
4                             41.0              4.0        0.097561
5                              9.0              0.0        0.000000

**Вывод:**
Самое первое, что бросается в глаза - процент просрочек у клиентов, у которых 5 детей. Так произошло по той причине, что у нас слишком маленькая выборка по такой категории:

In [25]:
data['children'].value_counts()

0    14091
1     4808
2     2052
3      330
4       41
5        9
Name: children, dtype: int64

По той же причине анализ клиентов, имеющих 3 и 4 детей может быть менее точным, но с другой стороны, такая выборка может быть обусловлена тем, что сейчас 1-2 ребёнка в семье - ситуация более распространённая, чем 3 и более.
Касательно итоговых результатов, с учётом малой выборки, видна тенденция, что чем больше детей в семье, тем чаще просрочки по кредиту. Это может быть обусловлено тем, что расходы семьи при наличии детей сложнее планировать:
- Дети растут и нужно постоянно покупать новую одежду и обувь
- Увеличиваются расходы на продукты питания
- Увеличиваются расходы на медицину
- И самое интересное для родителей - увеличиваются расходы на развлечения, досуг и игрушки

<div class="alert alert-success">
<b>Комментарий ревьюера №1</b> ✅
    
Хорошо
</div>

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [26]:
# ************ 1 вариант решения (до ревью) ************

# поймём, какие категории по есть по семейным статусам
family_statuses = sorted(data['family_status'].unique())
# print(family_statuses)
# Также есть поле family_status_id
family_status_ids = sorted(data['family_status_id'].unique())
# print(family_status_ids)
# Видим, что есть пять статусов и пять идентификаторов, при этом нет дубликатов

# Выведем информацию в виде сводной таблицы, где:
# Группировка будет по семейному положению: index=['family_status']
# Группируем значения по значениям задолженности: columns='debt'
# В качестве значений будет кол-во клиентов*: values = 'purpose'
# Кол-во клиентов находим через аггрегирующую функцию count: aggfunc='count'
data_pivot = data.pivot_table(index=['family_status'], columns='debt', values = 'purpose', aggfunc='count')
# print(data_pivot) # посмотрим, что получилось

# Переберём family_statuses и посчитаем процент заложенностей по каждому:
for family_status in family_statuses:
    result = data_pivot[1][family_status] / (data_pivot[1][family_status] + data_pivot[0][family_status])
    print(f'Семейное положение: {family_status}, процент просрочек: {result*100:.2f}%')

Семейное положение: Не женат / не замужем, процент просрочек: 9.76%
Семейное положение: в разводе, процент просрочек: 7.06%
Семейное положение: вдовец / вдова, процент просрочек: 6.62%
Семейное положение: гражданский брак, процент просрочек: 9.31%
Семейное положение: женат / замужем, процент просрочек: 7.56%


In [27]:
# ************ 2 вариант решения (после 1 ревью) ************

form_pivot_table(data, 'family_status')

Всего кредитополучателей  Всего должников  \
family_status                                                      
Не женат / не замужем                    2796.0            273.0   
в разводе                                1189.0             84.0   
вдовец / вдова                            951.0             63.0   
гражданский брак                         4134.0            385.0   
женат / замужем                         12261.0            927.0   

                       Доля должников  
family_status                          
Не женат / не замужем        0.097639  
в разводе                    0.070648  
вдовец / вдова               0.066246  
гражданский брак             0.093130  
женат / замужем              0.075606

**Вывод:** Аналогично с предыдущим заданием, оценим равномерность выборки:

In [28]:
data['family_status'].value_counts()

женат / замужем          12261
гражданский брак          4134
Не женат / не замужем     2796
в разводе                 1189
вдовец / вдова             951
Name: family_status, dtype: int64

~~При оценке по семейному положению выборка более равномерная, чем в предыдущем случае.~~

~~В результатах есть закономерность, выделяются два "лагеря":~~
~~1. Свободные люди (не женат / не замужем и гражданский брак (при учёте, если понимаем прод этим сожительство без оформления брака, а не классическое определение))~~
~~2. Семейные люди (женат / замужем, в разводе и вдовец / вдова)
Средний процент просрочек первой категории = 9.535%, второй = 7.08%, разница между ними = 2.455%~~

~~Почему так могло произойти. Как правило, люди, которые не создают в принципе (не женат / не замужем) или официально (гражданский брак) семью ещё молоды => могут быть проблемы со стабильной хорошо оплачиваемой работой, скорее всего, может отсутсвовать опыт ведения собственного бюджета или даже пробелы в финансовой грамотности.  
Люди семейные так или иначе сталкивались с вопросами планирования и распределения денег на некоторый период, поэтому им может быть легче с погашением в срок. Более того, как правило официально заводят семью люди, имеющие возможность её обеспечить финансово собственными силами, у них есть работа / сбережения / возможность "жить на зарплату одного, гасить задолженность с зарплаты другого".~~


<div class="alert alert-danger">
<b>Комментарий ревьюера №1</b> ❌
    
Среднее от среднего находить нельзя, оно будет отличаться от среднего без группировки. 
</div>

<div class="alert alert-info">
<b>Комментарий студента №2</b>
    
Убрала из вывода оценку среднего от среднего, в остальном вывод тот же
</div>

Вывод после 1 ревью:

При оценке по семейному положению выборка более равномерная, чем в предыдущем случае.

Как правило, люди, которые не создают в принципе (не женат / не замужем) или официально (гражданский брак) семью ещё молоды => могут быть проблемы со стабильной хорошо оплачиваемой работой, скорее всего, может отсутсвовать опыт ведения собственного бюджета или даже пробелы в финансовой грамотности.
Люди семейные так или иначе сталкивались с вопросами планирования и распределения денег на некоторый период, поэтому им может быть легче с погашением в срок. Более того, как правило официально заводят семью люди, имеющие возможность её обеспечить финансово собственными силами, у них есть работа / сбережения / возможность "жить на зарплату одного, гасить задолженность с зарплаты другого".



<div class="alert alert-success">
<b>Комментарий ревьюера №2</b> ✅
    
Люди, побывавшие в официальном браке, наиболее надежны. Возможно, есть связь с возрастом
</div>

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [29]:
# ************ 1 вариант решения (до ревью) ************

# Ранее, в 16 задаче, уже были разбиты по категориям уровни дохода, ими и воспользуемся
total_income_categories = sorted(data['total_income_category'].unique())
# print(total_income_categories)

# Выведем информацию в виде сводной таблицы, где:
# Группировка будет по категориям дохода: index=['total_income_category']
# Группируем значения по значениям задолженности: columns='debt'
# В качестве значений будет кол-во клиентов*: values = 'purpose'
# Кол-во клиентов находим через аггрегирующую функцию count: aggfunc='count'
data_pivot = data.pivot_table(index=['total_income_category'], columns='debt', values = 'purpose', aggfunc='count')
# print(data_pivot) # посмотрим, что получилось

# Переберём family_statuses и посчитаем процент заложенностей по каждому:
for total_income_category in total_income_categories:
    result = data_pivot[1][total_income_category] / (data_pivot[1][total_income_category] + data_pivot[0][total_income_category])
    print(f'Категория дохода: {total_income_category}, процент просрочек: {result*100:.2f}%')

Категория дохода: A, процент просрочек: 8.00%
Категория дохода: B, процент просрочек: 7.06%
Категория дохода: C, процент просрочек: 8.50%
Категория дохода: D, процент просрочек: 6.02%
Категория дохода: E, процент просрочек: 9.09%


In [30]:
# ************ 2 вариант решения (после 1 ревью) ************

form_pivot_table(data, 'total_income_category')

Всего кредитополучателей  Всего должников  \
total_income_category                                              
A                                          25.0              2.0   
B                                        5014.0            354.0   
C                                       15921.0           1353.0   
D                                         349.0             21.0   
E                                          22.0              2.0   

                       Доля должников  
total_income_category                  
A                            0.080000  
B                            0.070602  
C                            0.084982  
D                            0.060172  
E                            0.090909

**Вывод:** 
Ещё раз посмотрим вилки категорий:  
0–30000 — 'E';  
30001–50000 — 'D';  
50001–200000 — 'C';  
200001–1000000 — 'B';  
1000001 и выше — 'A'.  
Оценим равномерность выборки:

In [31]:
data['total_income_category'].value_counts()

C    15921
B     5014
D      349
A       25
E       22
Name: total_income_category, dtype: int64

Видим, что категориии A и E, как крайние, проседают относительно "середины", поэтому тут может быть погрешность. Например, странно, что категория B имеет меньший процент просрочек, чем категория A (отрыв в 0.94%), скорее всего, в действительности, клиентам из категории А проще закрыть кредит вовремя.  
Дальше рассмотрим категорию E, процент просрочек у них наиболее высокий, отсюда можем сделать вывод, что если клиент получает 30 тысяч и меньше, получаем наибольшую вероятность просрочки.  
Категории C и D наиболее близки к медианным значениям, но между их выборками тоже (как у A и B) большая разница, поэтому результат по категории D не соответствует реальности. Предположу, что фактическое значение на большей выборки будет где-то между 9.09% и 8.5%.

Итого, получается, что, чем выше доход клиента, тем с большей вероятностью кредит будет погашен в срок.


<div class="alert alert-success">
<b>Комментарий ревьюера №1</b> ✅
    
Здесь лучше делать выводы по двум самым крупным категориям - B и С. По остальным категориям данных недостаточно для корректных выводов, статпогрешность большая. Чем выше доход, тем ниже доля должников
</div>

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [32]:
# ************ 1 вариант решения (до ревью) ************

# Ранее, в 18 задаче, уже были разбиты по категориям цели взятия кредита, ими и воспользуемся
purpose_categories = sorted(data['purpose_category'].unique())
# print(purpose_categories)

# Выведем информацию в виде сводной таблицы, где:
# Группировка будет по категориям цели взятия кредита: index=['purpose_category']
# Группируем значения по значениям задолженности: columns='debt'
# В качестве значений будет кол-во клиентов*: values = 'purpose'
# Кол-во клиентов находим через аггрегирующую функцию count: aggfunc='count'
data_pivot = data.pivot_table(index=['purpose_category'], columns='debt', values = 'purpose', aggfunc='count')
# print(data_pivot) # посмотрим, что получилось

# Переберём purpose_categories и посчитаем процент заложенностей по каждому:
for purpose_category in purpose_categories:
    result = data_pivot[1][purpose_category] / (data_pivot[1][purpose_category] + data_pivot[0][purpose_category])
    print(f'Цель взятия кредита: {purpose_category}, процент просрочек: {result*100:.2f}%')

Цель взятия кредита: операции с автомобилем, процент просрочек: 9.35%
Цель взятия кредита: операции с недвижимостью, процент просрочек: 7.26%
Цель взятия кредита: получение образования, процент просрочек: 9.25%
Цель взятия кредита: проведение свадьбы, процент просрочек: 7.91%


In [33]:
# ************ 2 вариант решения (после 1 ревью) ************

form_pivot_table(data, 'purpose_category')

Всего кредитополучателей  Всего должников  \
purpose_category                                                      
операции с автомобилем                      4279.0            400.0   
операции с недвижимостью                   10751.0            780.0   
получение образования                       3988.0            369.0   
проведение свадьбы                          2313.0            183.0   

                          Доля должников  
purpose_category                          
операции с автомобилем          0.093480  
операции с недвижимостью        0.072551  
получение образования           0.092528  
проведение свадьбы              0.079118

**Вывод:** Оценим равномерность выборки:

In [34]:
data['purpose_category'].value_counts()

операции с недвижимостью    10751
операции с автомобилем       4279
получение образования        3988
проведение свадьбы           2313
Name: purpose_category, dtype: int64

Выборка достаточно равномерна, более того, хорошо видно, что чаще всего кредиты берут на покупку недвижимости.

Начнём с недвижимости. Как правило, покупка недвижимости - это очень крупная сделка, к которой нужно подготовиться морально и финансово, иначе её можно и не инициировать. Ввиду того, что такие кредиты берутся на десятилетия, клиенты очень ответственно к ним подходят, т.к. не хотят потерять жильё за непогашение кредита. Такие кредиты просрочиваются из-за непредвиденных обстоятельств на длинной дистанции (несчатный случай, сокращение и т.п.).  
Следующая цель - проведение свадьбы. Специфика этого мероприятия - небольшая длительность по времени (как само событие, это чаще всего 1 день) относительно других целей и некоторый "кешбек" в виде подарков от гостей. В среднем, на организацию свадьбы тратят ~300 тысяч, что делает этот кредит ещё и самым маленьким по стоимости. И эти три фактора делают кредит на проведение свадьбы вторым по безопасности для банка.  
Дальше с отрывом более чем в процент идут категории операции с автомобилем и получение образования. Продолжим с образованием. Средняя стоимость обучения на программах бакалавриата и специалитета на сегодняшний день составляет 184.5 тыс. рублей в год, это в среднем 738-922.5 тыс. рублей за весь период обучения. Если перевести на ежемесячный платёж - 14000. Эти цифры были без банковских процентов. Если обучение оплачивают родители, не каждая семья может себе позволить такие дополниельные расходы. Если же сами студенты, пытаясь совместить работу и учёбу, тут стоит уточнить, что на неполную ставку и без оконченного высшего или специального образования большие зарплаты не предлагают, поэтому, сумма платежа тоже не всем будт по карману.  
Что с автомобилями, последнее время цены на авто сильно выросли, часть марок ушла с рынка и вместе с эти увеличился ассортимент китайских автомобилей => выбор, как и оценку фактической стоимости, стало сделать сложнее, клиенты вынуждены брать кредиты на большую сумму, чем они могли бы до изменений на рынке. Если принять во внимание качество материалов и сборки бюджетных и средних по стоимости авто, то клиент будет нести дополнительные расходы (не всегда ожидаемые) на ремонт.

Безопаснее всего давать кредиты на покупку недвижимости или проведение свадьбы.

<div class="alert alert-success">
<b>Комментарий ревьюера №1</b> ✅ 
    
Верно
</div>

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:*  
Пропуски могут возникнуть по нескольким причинам:
1. Сбои в работе/операциях БД
2. Ошибки при ручном заполнении оператором
3. Ошибки при ручном заполнением через онлайн анкету на стороне web-сервиса
4. Аппаратные проблемы

<div class="alert alert-success">
<b>Комментарий ревьюера №1</b> ✅
    
Клиенты могли не предоставить данные. Возможно, эту информацию не требуют в некоторых случаях. Проблемы могли возникнуть на этапе загрузки и на этапе выгрузки данных. Неправильный запрос, объединение таблиц также может быть причиной.
</div>

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:*  
Пропуски для количественных переменных можно обработать несколькими способами:
- Оставить их как есть => получаем искажение в итоговых результатах
- Удалить строки, где есть пропущенных значения => теряем часть выборки
- Заполнить средним арифметическим значением => не плохой вариант, если нет артефактов (очень больших или очень маленьких значений относительно общей массы), если они есть, то заполнение будет не наиболее точным из возможного
- Заполнить медианным значением => оптимальный вариант, которому не страшны артефакты

И таким образом, медианное значение - наилучший вариант обработки пропусков.

<div class="alert alert-success">
<b>Комментарий ревьюера №1</b> ✅
    
Да, смотрим на наличие выбросов👍
</div>

### Шаг 4: общий вывод.

Для получения полезной информации из пула данных, необходимо провести эти данные через этап предобработки:
1. Изучить имеющиеся данные
2. Обработать пропуски
3. Обработать явные и не явные дубликаты
4. Категоризовать данные 

После того, как данные обработаны, можно переходить к их анализу для того, чтобы получить полезную для бизнеса информацию. В рамках исследования надёжности заёмщиков можем сделать ряд заключений по категориям заёмщиков, кто реже опаздывает с погашением кредита:
1. "Семейные" клиенты
2. Клиенты, чей доход начинается от 50001 рублей
3. Клиенты, берущие кредит на покупку недвижимости или проведение свадьбы
4. Клиенты без детей

<div class="alert alert-danger">
<b>Комментарий ревьюера №1</b> ❌

Выводы по доходу нужно будет исправить. Опиши все зависимости. Здесь также нужно добавить рекомендации банку-заказчику

</div>

<div class="alert alert-info">
<b>Комментарий студента №3</b>
    
Ниже добавила дополнение по общему выводу с недостающей информацией. Если вывод всё же не тот, можно переформулировать часть комментария "Выводы по доходу нужно будет Опиши все зависимости"?
</div>

Дополнения после 1 ревью:  

**Подробнее про доходы клиента.**  
Относительно точные показатели возможно дать только по клиентам, доходы которых относятся к категориям:  
50001–200000 руб. — 'C'  
200001–1000000 руб. — 'B'  
Т.к. по остальным очень мало данных; кол-во клинетов по категориям их дохода:  
C = 15921 чел.  
B = 5014 чел.  
D = 349 чел.  
A = 25 чел.  
E = 22 чел.  

Исходя из того, что валидными для анализа остаются две категории:  
Категория дохода: B, процент просрочек: 7.06%  
Категория дохода: C, процент просрочек: 8.50%  
Между ними есть зависимость: чем выше доход клиента, тем больше вероятность, что он не просрочит платёж.

**Рекомендации банку-заказчику.**  
Исходя из анализа имеющихся данных, можем выделить ряд рекомендаций, которые помогут снизить вероятность просрочек платежей банку-заказчику по кредитам:  
1. Чем выше доход клиента, тем с большей вероятностью кредит будет погашен в срок
2. Бездетные клиенты реже просрачивают платежи
3. По ипотекам и кредитам на свадьбу просрочки реже, чем по кредитам на авто и образование
4. Клиенты состоящие или состоявшие в браке отвественнее подходят к погашению платежей в срок

**Портрет идеального клиента**, запрашивающего кредит - бездетный женатый/замужний клиент, имеющий доход категории B и выше, берущий ипотеку или кредит на свадьбу.  
Т.к. клиентов, у которых доход начинается от 200 тыс. руб., меньше, чем клиентов среднего достатка (50-200 тыс. руб.), банк-заказчик потеряет много возможных сделок, поэтому одобрять кредиты только для "идеального клиента" не выгодно. В вопросе дохода клиента можно ориентироваться на общую рекомендацию №1 (из списка выше).


<div class="alert alert-success">
<b>Комментарий ревьюера №2</b> ✅
    
В качестве рекомендаций можно предложить другую категоризацию по доходам. Или исследовать другие признаки, рассчитать другие метрики (отношение дохода к сумме выплат и др.). В анкете сделать поля о доходе и стаже обязательными к заполнению, добавить автопроверку на формат вводимых пользователем данных
</div>

<div style="border:solid purple 2px; padding: 20px"> 
<b> <font color='purple' >   
    
В целом, работа хорошая. У тебя отличные аналитические навыки, но остались небольшие доработки. Нужно вывести сводные в каждом вопросе, проверить выводы. Доделывай и присылай на повторную проверку.
        
Жду твоих исправлений 😊
</font></b>
</div>

<div style="border:solid purple 2px; padding: 20px"> 
<b> <font color='purple' >   
Отлично! Критических замечаний больше нет. Проект принят! 👏 Поздравляю и желаю удачи!
</font></b>
</div>